SIPS/sip-27.md

```diff
@@ -0,0 +1,64 @@
+---
+sip: 27
+title: Accounts Metadata Request
+status: Draft
+author: Olaf Tomalka (@ritave)
+created: 2024-09-17
+---
+
+## Abstract
+
+This SIP allows snaps to retrieve metadata related to accounts that exist in the extension.
+
+## Motivation
+
+The intention of this SIP is to allow snaps providing new accounts for new chains to be able to list all accounts when selecting one during transfers.
+
+## Specification
+
+> Indented sections like this are considered non-normative.
+
+> Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs).
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies a permission named `keyring_listAccountsAll`. This permission grants a snap the ability to retrieve account metadata through an RPC call.
+
+The permission is specified in `snap.manifest.json` as follows:
+
+```json
+{
+  "initialPermissions": {
+    "keyring_listAccountsAll": {}
+  }
+}
+```
+
+### RPC Method
+
+This SIP exposes a new RPC method called `keyring_listAccountsAll` with no additional parameters.
+
+The RPC call returns with the following data:
+
+```typescript
+type KeyringAccount = {
+  id: string; // An extension-specific unique ID
+  address: string; // A blockchain specific public address for the account.
```

Comment:
Should this be a CAIP-10 address?

Reply:
This address can exist on multiple chains at the same time. Keyrings initially had a `chains` property, but it's not implemented

Reply:
Then how would you identify which network the address is tied to? 🤔

Reply:
The only data available in MetaMask currently is the `type` property.

Reply:
I guess my question is, should we combine `type` and `address` into one field with a CAIP-10 address.

Reply:
`type` exposes more information than just the network the address is on.

Accounts team envisioned the following types:
```typescript
type AccountType =
	| 'eip155:eoa'
	| 'eip155:sca:erc4337'
	| 'bip122:p2pk'
	| 'bip122:p2pkh'
	| 'bip122:p2sh'
	| 'bip122:p2wpkh'
	| 'bip122:p2tr';
```.
Notice they are all on mainnet while still differing

Reply:
The main goal of `type` is to differentiate between accounts that may behave differently, such as EOAs and Smart Contract Accounts.

For multichain, we are still working on a solution that addresses two key issues:
- Storing past Bitcoin addresses
- Supporting multiple addresses for multiple chains (e.g., Cosmos)

We are considering approaches like `addresses: Map<Caip2ChainId, Address[]>` or `addresses: Caip10Address[]` [(PR draft](https://github.com/MetaMask/keyring-api/pull/343/files)). Which is aligned with what you said above.

Reply:
@ritave, I need to double-check, but the reason we dropped the chains field was that CAIP-2 doesn't support wildcards. We considered using `eip155:*`, even though it wouldn’t be compliant, but ultimately decided to replace it with a `filterAccountChains` method. This method can be called with a list of CAIP-2 chain IDs, and the Snap will return the list of supported chains from the input where the account can be used.

I'm not super happy with the decision and think we can reconsided the `chains` field.

Reply:
IIRC, we were also considering using just the CAIP namespace as an alternative to wildcard. This way we can regroup "things" on the same namespace, so instead of `eip155:*`, we would use `eip155`.

We started using this pattern in the extension/core IIRC. But this is still not 100% compliant with a true CAIP-2 identifier...